huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
         self.logger.debug("Flat-field observation: {}".format(flat_obs))
 
         return flat_obs
+
+    def _take_autoexposure_flats(self, cameras, observation, min_counts=5000, max_counts=15000,
+                                 bias=1000, max_exptime=60., target_adu_percentage=0.5,
+                                 max_num_exposures=10, filter_name=None, min_saturated_exptime=2,
+                                 min_exptime=1, *args, **kwargs):
+        """Take flat fields iteratively by automatically estimating exposure times.
+
+        Args:
+            cameras (dict): Dict of camera name: Camera pairs.
+            filter_names (dict): Dict of filter name for each camera.
+        """
+        target_adu = target_adu_percentage * (min_counts + max_counts)
+        exptimes = {cam_name: [1. * u.second] for cam_name in cameras.keys()}
+
+        # Loop until conditions are met to finish flat-fielding
+        exposure_count = 0
+        while True:
+
+            # Update the exposure count
+            exposure_count += 1
+
+            # Get the FITS headers with a common start time
+            start_time = utils.current_time()
+            fits_headers = self.get_standard_headers(observation=observation)
+            fits_headers['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            next_exptimes = {c: exptimes[c][-1] for c in cameras.keys()}
+            camera_events = self._take_flat_fields(next_exptimes, observation,
+                                                   fits_headers=fits_headers)
+
+            # Check the counts for each image
+            is_saturated = False
+            for cam_name, info in camera_events.items():
+                filename = info['filename']
+                self.logger.debug("Checking counts for {}".format(filename))
+
+                # Load the image we just took
+                try:
+                    data = fits.getdata(filename)
+                except FileNotFoundError:
+                    data = fits.getdata(filename + '.fz')
+
+                # Calculate average counts per pixel
+                mean_counts, _, _ = stats.sigma_clipped_stats(data)
+                mean_counts = mean_counts - bias
+                self.logger.debug("Counts: {}".format(mean_counts))
+                if (mean_counts < min_counts) or (mean_counts > max_counts):
+                    self.logger.debug("Counts outside min/max range, should be discarded.")
```

Comment:
Remove increment of `exposure_count` from line 400 and do it here, but only if the counts are good. We should be using the number of good flat field exposures as the condition to stop taking flats, not just the number of attempted flat field exposures.

```suggestion
                if (mean_counts < min_counts) or (mean_counts > max_counts):
                    self.logger.debug("Counts outside min/max range, should be discarded.")
                else:
                    exposure_count += 1
```

Reply:
Looking at the code, `exposure_count` is not actually used anywhere. Instead, this behaviour is currently controlled by incrementing `max_num_exposures`.

Reply:
The other problem with doing the increment where you suggested is that its inside the loop over cameras.

Reply:
Yes, I realised later that this was in the camera loop so wouldn't work here. I think it's right that we want to do this so that we continue with flats for a given filter until all the cameras have a certain minimum number of 'good' flats (i.e. between `min_counts` and `max_counts`) but the implementation is a bit awkward.

I don't like incrementing `max_num_exposures` as a solution, seems confusing to me moving the goal posts like that... The alternative of decrementing `exposure_count` when finding a bad exposure is pretty ugly too, but I think that would be more in keeping with the meaning of the two variables.